src/combinatorics/colex.lean

```diff
@@ -0,0 +1,236 @@
+instance [linear_order α] : is_trans (finset α) (<ᶜ) :=
+begin
+  constructor,
+  rintros A B C ⟨k₁, k₁z, notinA, inB⟩ ⟨k₂, k₂z, notinB, inC⟩,
+  have: k₁ ≠ k₂ := ne_of_mem_of_not_mem inB notinB,
+  cases lt_or_gt_of_ne this,
+    refine ⟨k₂, _, by rwa k₁z h, inC⟩,
+    intros x hx, rw ← k₂z hx, apply k₁z (trans h hx),
+  refine ⟨k₁, _, notinA, by rwa ← k₂z h⟩,
+  intros x hx, rw k₁z hx, apply k₂z (trans h hx)
+end
+instance [linear_order α] : is_asymm (finset α) (<ᶜ) := by apply_instance
+instance [linear_order α] : is_antisymm (finset α) (≤ᶜ) :=
+⟨λ A B AB BA, AB.elim (λ k, BA.elim (λ t, (asymm k t).elim) (λ t, t.symm)) id⟩
+instance [linear_order α] : is_trans (finset α) (≤ᶜ) :=
+⟨λ A B C AB BC, AB.elim (λ k, BC.elim (λ t, or.inl (trans k t)) (λ t, t ▸ AB))
+                        (λ k, k.symm ▸ BC)⟩
+instance [linear_order α] : is_strict_order (finset α) (<ᶜ) := {}
+
+instance [linear_order α] : is_trichotomous (finset α) (<ᶜ) :=
+begin
+  split, intros A B,
+  classical,
+  by_cases (A = B), right, left, assumption,
+  rcases (exists_max_image (A \ B ∪ B \ A) id _) with ⟨k, hk, z⟩,
+    simp at hk, cases hk, right, right, swap, left, swap,
+      any_goals { refine ⟨k, λ t th, _, hk.2, hk.1⟩, specialize z t, by_contra,
+                  simp only [mem_union, mem_sdiff, id] at z, rw [not_iff,
+                  iff_iff_and_or_not_and_not, not_not, and_comm] at a,
+                  apply not_le_of_lt th (z _) },
+      { exact a }, { exact a.symm },
+  rw nonempty_iff_ne_empty,
+  intro a, simp only [union_eq_empty_iff, sdiff_eq_empty_iff_subset] at a,
+  apply h (subset.antisymm a.1 a.2)
+end
+
+instance [linear_order α] : is_total (finset α) (≤ᶜ) := ⟨λ A B,
+(trichotomous A B).elim3 (or.inl ∘ or.inl) (or.inl ∘ or.inr) (or.inr ∘ or.inl)⟩
+
+instance [linear_order α] :
+  is_linear_order (finset α) (≤ᶜ) := {}
+instance [linear_order α] : is_incomp_trans (finset α) (<ᶜ) :=
+begin
+  constructor,
+  rintros A B C ⟨nAB, nBA⟩ ⟨nBC, nCB⟩,
+  have: A = B := ((trichotomous A B).resolve_left nAB).resolve_right nBA,
+  have: B = C := ((trichotomous B C).resolve_left nBC).resolve_right nCB,
+  rw [‹A = B›, ‹B = C›], rw and_self, apply irrefl
+end
+instance [linear_order α] :
+  is_strict_weak_order (finset α) (<ᶜ) := {}
+instance [linear_order α] :
+  is_strict_total_order (finset α) (<ᶜ) := {}
+instance colex_order [has_lt α] : has_le (finset α) := {le := (≤ᶜ)}
+instance colex_preorder [linear_order α] : preorder (finset α) :=
+{le_refl := refl_of (≤ᶜ), le_trans := is_trans.trans, ..colex_order}
+instance colex_partial_order [linear_order α] : partial_order (finset α) :=
+{le_antisymm := is_antisymm.antisymm, ..colex_preorder}
+instance colex_linear_order [linear_order α] :
+  linear_order (finset α) :=
+{le_total := is_total.total, ..colex_partial_order}
+
+/--
+Rewrite colex in a particular way so that it uses only bounded quantification,
+so we can infer decidability.
+-/
+lemma colex_dec [has_lt α] (A B : finset α) : A <ᶜ B ↔
```

Comment:
The stuff here can probably go - they let us infer decidability but this can just be done by going classical, so this isn't really necessary